p2p/net/network.go

```diff
@@ -124,10 +123,81 @@ func (n *netImpl) DialTCP(address string, timeOut time.Duration, keepAlive time.
 	}
 
 	n.logger.Debug("Connected to %s...", address)
-	c := newConnection(netConn, n, Local)
+	c := newConnection(netConn, n, Local, remotePub, n.logger)
+
 	return c, nil
 }
 
+func (n *netImpl) createSecuredConnection(address string, remotePublicKey crypto.PublicKey, networkId int8, timeOut time.Duration, keepAlive time.Duration) (Connectioner, error) {
+	errMsg := "failed to establish secured connection."
+	conn, err := n.createConnection(address, remotePublicKey, timeOut, keepAlive)
+	if err != nil {
+		return nil, err
+	}
+	data, session, err := GenerateHandshakeRequestData(n.localNode.PublicKey(), n.localNode.PrivateKey(), remotePublicKey, networkId)
+	if err != nil {
+		conn.Close()
+		return nil, fmt.Errorf("%s err: %v", errMsg, err)
+	}
+	n.logger.Debug("Creating session handshake request session id: %s", session)
+	payload, err := proto.Marshal(data)
+	if err != nil {
+		conn.Close()
+		return nil, fmt.Errorf("%s err: %v", errMsg, err)
+	}
+
+	err = conn.Send(payload)
+	if err != nil {
+		return nil, err
+	}
+
+	var msg wire.InMessage
+	var ok bool
+	timer := time.NewTimer(n.config.ResponseTimeout)
+	select {
+	case msg, ok = <-conn.IncomingChannel():
+		if !ok {
+			return nil, fmt.Errorf("%s err: incoming channel got closed", errMsg)
+		}
+		if msg.Error() != nil {
+			return nil, fmt.Errorf("%s err: %v", errMsg, msg.Error())
+		}
+	case <-timer.C:
+		n.logger.Info("waiting for HS response timed-out. remoteKey=%v", remotePublicKey)
+		return nil, fmt.Errorf("%s err: HS response timed-out", errMsg)
+	}
+
+	respData := &pb.HandshakeData{}
+	err = proto.Unmarshal(msg.Message(), respData)
+	if err != nil {
+		//n.logger.Warning("invalid incoming handshake resp bin data", err)
+		conn.Close()
+		return nil, fmt.Errorf("%s err: %v", errMsg, err)
+	}
+
+	err = ProcessHandshakeResponse(remotePublicKey, session, respData)
+	if err != nil {
+		conn.Close()
+		return nil, fmt.Errorf("%s err: %v", errMsg, err)
+	}
+
+	conn.SetSession(session)
+	return conn, nil
+}
+
+// Dial a remote server with provided time out
+// address:: ip:port
+// Returns established connection that local clients can send messages to or error if failed
+// to establish a connection, currently only secured connections are supported
+func (n *netImpl) Dial(address string, remotePublicKey crypto.PublicKey, networkId int8) (Connectioner, error) {
+	conn, err := n.createSecuredConnection(address, remotePublicKey, networkId, n.config.DialTimeout, n.config.ConnKeepAlive)
+	if err != nil {
+		return nil, fmt.Errorf("failed to Dail. err: %v", err)
+	}
+	go conn.beginEventProcessing()
```

Comment:
since this method is called beginEventProcessing and i see no reason why it should ever run in blocking mode consider moving the goroutine inside beginEventProcessing itself